cpu/stm32f0/periph/gpio.c

```diff
@@ -341,209 +152,77 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pullup, gpio_flank_t flank, gpio_cb_t cb
     return 0;
 }
 
-void gpio_irq_enable(gpio_t dev)
+void gpio_init_af(gpio_t pin, gpio_af_t af)
 {
-    uint8_t pin;
-
-    if (dev >= GPIO_NUMOF) {
-        return;
-    }
-
-    pin = gpio_pin_map[dev];
-    EXTI->IMR |= (1 << pin);
+    GPIO_TypeDef *port = _port(pin);
+    uint32_t pin_num = _pin_num(pin);
+
+    /* set pin to AF mode */
+    port->MODER &= ~(3 << (2 * pin_num));
+    port->MODER |= (2 << (2 * pin_num));
+    /* set selected function */
+    port->AFR[(pin_num > 7) ? 1 : 0] &= ~(0xf << ((pin_num & 0x07) * 4));
+    port->AFR[(pin_num > 7) ? 1 : 0] |= (af << ((pin_num & 0x07) * 4));
 }
 
-void gpio_irq_disable(gpio_t dev)
+void gpio_irq_enable(gpio_t pin)
 {
-    uint8_t pin;
-
-    if (dev >= GPIO_NUMOF) {
-        return;
-    }
-
-    pin = gpio_pin_map[dev];
-    EXTI->IMR &= ~(1 << pin);
+    EXTI->IMR |= (1 << _pin_num(pin));
 }
 
-int gpio_read(gpio_t dev)
+void gpio_irq_disable(gpio_t pin)
 {
-    GPIO_TypeDef *port;
-    uint8_t pin;
-
-    if (dev >= GPIO_NUMOF) {
-        return -1;
-    }
+    EXTI->IMR &= ~(1 << _pin_num(pin));
+}
 
-    port = gpio_port_map[dev];
-    pin = gpio_pin_map[dev];
+int gpio_read(gpio_t pin)
+{
+    GPIO_TypeDef *port = _port(pin);
+    uint32_t pin_num = _pin_num(pin);
 
-    if (port->MODER & (1 << (pin * 2))) {       /* if configured as output */
-        return port->ODR & (1 << pin);          /* read output data register */
+    if (port->MODER & (1 << (pin_num * 2))) {   /* if configured as output */
+        return port->ODR & (1 << pin_num);      /* read output data reg */
     } else {
-        return port->IDR & (1 << pin);          /* else read input data register */
+        return port->IDR & (1 << pin_num);      /* else read input data reg */
     }
 }
 
-void gpio_set(gpio_t dev)
+void gpio_set(gpio_t pin)
 {
-    GPIO_TypeDef *port;
-    uint8_t pin;
-
-    if (dev >= GPIO_NUMOF) {
-        return;
-    }
-
-    port = gpio_port_map[dev];
-    pin = gpio_pin_map[dev];
-
-    port->ODR |= (1 << pin);
+    _port(pin)->BSRRL = (1 << _pin_num(pin));
 }
 
-void gpio_clear(gpio_t dev)
+void gpio_clear(gpio_t pin)
 {
-    GPIO_TypeDef *port;
-    uint8_t pin;
-
-    if (dev >= GPIO_NUMOF) {
-        return;
-    }
-
-    port = gpio_port_map[dev];
-    pin = gpio_pin_map[dev];
-
-    port->ODR &= ~(1 << pin);
+    _port(pin)->BSRRH = (1 << _pin_num(pin));
 }
 
-void gpio_toggle(gpio_t dev)
+void gpio_toggle(gpio_t pin)
 {
-    if (gpio_read(dev)) {
-        gpio_clear(dev);
+    if (gpio_read(pin)) {
+        gpio_clear(pin);
     } else {
-        gpio_set(dev);
+        gpio_set(pin);
     }
 }
 
-void gpio_write(gpio_t dev, int value)
+void gpio_write(gpio_t pin, int value)
 {
     if (value) {
-        gpio_set(dev);
+        gpio_set(pin);
```

Comment:
you could gain slightly more efficiency by saving the function call here: 

s/gpio_set(pin);/_port(pin)->BSRRL = (1 << _pin_num(pin));/